sqlalchemy-stubs/engine/base.pyi

```diff
@@ -1,138 +1,172 @@
-    parameters: Any = ...
-    is_disconnect: Any = ...
-    invalidate_pool_on_disconnect: Any = ...
+    engine: Engine = ...
+    connection: Connection = ...
+    sqlalchemy_exception: Optional[StatementError] = ...
+    original_exception: BaseException = ...
+    execution_context: Optional[ExecutionContext] = ...
+    statement: Optional[str] = ...
+    parameters: Optional[Any] = ...
+    is_disconnect: bool = ...
+    invalidate_pool_on_disconnect: bool = ...
     def __init__(
         self,
-        exception: Any,
-        sqlalchemy_exception: Any,
-        engine: Any,
-        connection: Any,
-        cursor: Any,
-        statement: Any,
-        parameters: Any,
-        context: Any,
-        is_disconnect: Any,
-        invalidate_pool_on_disconnect: Any,
+        exception: BaseException,
+        sqlalchemy_exception: Optional[StatementError],
+        engine: Engine,
+        connection: Optional[Connection],
+        cursor: Optional[_DBAPICursor],
+        statement: Optional[str],
+        parameters: Optional[Any],
+        context: Optional[ExecutionContext],
+        is_disconnect: bool,
+        invalidate_pool_on_disconnect: bool,
     ) -> None: ...
 
 class Transaction:
-    def __init__(self, connection: Any) -> None: ...
+    def __init__(self, connection: Connection) -> None: ...
     @property
-    def is_valid(self): ...
+    def is_valid(self) -> bool: ...
     def close(self) -> None: ...
     def rollback(self) -> None: ...
     def commit(self) -> None: ...
-    def __enter__(self): ...
+    def __enter__(self: _TTransaction) -> _TTransaction: ...
     def __exit__(self, type_: Any, value: Any, traceback: Any) -> None: ...
 
 class MarkerTransaction(Transaction):
-    connection: Any = ...
-    def __init__(self, connection: Any) -> None: ...
+    connection: Connection = ...
+    def __init__(self, connection: Connection) -> None: ...
     @property
-    def is_active(self): ...
+    def is_active(self) -> bool: ...
 
 class RootTransaction(Transaction):
-    connection: Any = ...
+    connection: Connection = ...
     is_active: bool = ...
-    def __init__(self, connection: Any) -> None: ...
+    def __init__(self, connection: Connection) -> None: ...
 
 class NestedTransaction(Transaction):
-    connection: Any = ...
+    connection: Connection = ...
     is_active: bool = ...
-    def __init__(self, connection: Any) -> None: ...
+    def __init__(self, connection: Connection) -> None: ...
 
 class TwoPhaseTransaction(RootTransaction):
-    xid: Any = ...
-    def __init__(self, connection: Any, xid: Any) -> None: ...
+    xid: str = ...
+    def __init__(self, connection: Connection, xid: str) -> None: ...
     def prepare(self) -> None: ...
 
 class Engine(Connectable, log.Identified):
-    pool: Any = ...
+    pool: Pool = ...
     url: Any = ...
-    dialect: Any = ...
-    logging_name: Any = ...
+    dialect: Dialect = ...
+    logging_name: Optional[str] = ...
     echo: Any = ...
```

Comment:
None | bool | str (iirc the string can only be 'debug')  see https://github.com/sqlalchemy/sqlalchemy/blob/a323b16507a76f7e8db857ab9449c66691802325/lib/sqlalchemy/log.py#L91-L94

this is also fine as is